locator/topology.cc

```diff
@@ -16,6 +16,32 @@
 #include "locator/production_snitch_base.hh"
 #include "utils/stall_free.hh"
 
+struct node_printer {
+    const locator::node* v;
+    node_printer(const locator::node* n) noexcept : v(n) {}
+};
+
+template <>
+struct fmt::formatter<node_printer> {
+    constexpr auto parse(format_parse_context& ctx) { return ctx.begin(); }
+    auto format(const node_printer& np, fmt::format_context& ctx) const {
+        const locator::node* node = np.v;
+        auto out = fmt::format_to(ctx.out(), "node={}", fmt::ptr(node));
+        if (node) {
+            out = fmt::format_to(out, " idx={} host_id={} endpoint={} dc={} rack={} state={} shards={} this_node={}",
```

Comment:
how about taking this opportunity to improve the existing `fmt::formatter<locator::node>`? so we can reuse it here?

Reply:
to be more specific, i am proposing something like:
```c++
# {:d} for debug format which is more verbose
template <>
struct fmt::formatter<locator::node> : fmt::formatter<std::string_view> {
    bool verbose = false;
    constexpr auto parse(fmt::format_parse_context& ctx) {
        auto it = ctx.begin();
        auto end = ctx.end();
        if (it != end) {
            if (*it == 'd') {
                _verbose = true;
                ++it;
            }
            if (it != end && *it != '}') {
                throw fmt::format_error("invalid format specifier");
            }
        }
        return it;
    }
    template <typename FormatContext>
    auto format(const locator::node& node, FormatContext& ctx) const {
        auto out = ctx.out();
        if (verbose) {
            // print more fields
        } else {
            out = fmt::format_to(out, "{}/{}", node.host_id(), node.endpoint());
        }
        return out;
    }
};
```

Reply:
nuce idea!
1. none if that runs if the logging level is lower rhan the statement's, right?
2. should we also ignore `:` in `parse`?

Reply:
Also, I guess wecwould like to reuse and formalize the flags, so I suggest implementing the parser in e.g utils/log.hh
Then formatters can get the flags as enum set and decide what to do about them.

If we pass the utils library parser an enum set of the flags the parser supports, we can also do the error handling for ubsupported flags there.

Reply:
> nuce idea!
> 
>     1. none if that runs if the logging level is lower rhan the statement's, right?

right. that's exactly a part Pavel's change's intention: to make the formatting lazy.

> 
>     2. should we also ignore `:` in `parse`?

no, we should not. as ':' marks the beginning of format specifiers which is passed to `parse()`. see https://fmt.dev/latest/syntax.html#format-string-syntax . and i am quoting the related paragraphs here

> replacement_field ::=  "{" [[arg_id](https://fmt.dev/latest/syntax.html#grammar-token-sf-arg_id)] [":" ([format_spec](https://fmt.dev/latest/syntax.html#grammar-token-sf-format_spec) | [chrono_format_spec](https://fmt.dev/latest/syntax.html#grammar-token-sf-chrono_format_spec))] "}"

Reply:
Good point. 

For the record: in #14122 there's also an idea to print extended info about sstables with custom format specifiers